### PR TITLE
Feature/csl937 body hash to header

### PR DESCRIPTION
**Pull request by 7sedam7**
CSL-937 Add `Hash MainExtraBodyData` to `MainExtraHeaderData`

### CODE REVIEW

File: src/Pos/Block/Core/Main/Types.hs [hunk 1]
```diff
@@ -76,6 +76,7 @@ data MainExtraHeaderData = MainExtraHeaderData
       _mehSoftwareVersion :: !SoftwareVersion
     , -- | Header attributes
       _mehAttributes      :: !BlockHeaderAttributes
+    , _mehEBDataProof     :: !(Hash MainExtraBodyData)
```
Comment (gromakovsky) [inline src/Pos/Block/Core/Main/Types.hs]: Can you add a haddock comment for consistency?

File: src/Pos/Block/Logic/Creation.hs [hunk 2]
```diff
@@ -46,7 +46,8 @@ import           Pos.Core                   (Blockchain (..), EpochIndex,
                                              SlotId (..), SlotLeaders, crucialSlot,
                                              epochOrSlot, flattenSlotId, getEpochOrSlot,
                                              getSlotIndex, headerHash, mkLocalSlotIndex)
-import           Pos.Crypto                 (SecretKey, WithHash (WithHash))
+import           Pos.Crypto                 (SecretKey, WithHash (WithHash), Hash)
+import           Pos.Crypto.Hashing         (hash)
```
Comment (gromakovsky) [inline src/Pos/Block/Logic/Creation.hs]: I think `hash` can also be imported from `Pos.Crypto` (without `Hashing`), but it's not really important.

Comment (georgeee) [inline src/Pos/Block/Logic/Creation.hs]: Right, it can be

**Comment by 7sedam7** [comment]
I don't know should I write here when changes are done, but I see that nobody commented on the changes so @gromakovsky

File: src/Pos/Block/Core/Main/Types.hs [hunk 3]
```diff
@@ -76,7 +76,8 @@ data MainExtraHeaderData = MainExtraHeaderData
       _mehSoftwareVersion :: !SoftwareVersion
     , -- | Header attributes
       _mehAttributes      :: !BlockHeaderAttributes
-    , _mehEBDataProof     :: !(Hash MainExtraBodyData)
+    , -- | Body Hash
```
Comment (gromakovsky) [inline src/Pos/Block/Core/Main/Types.hs]: It's not Body hash, it's hash of extra data, which is different from Body.

**Comment by gromakovsky** [comment]
@7sedam7 I saw your commit, but you didn't add check that hash is correct as I wrote above:
> You should also check that hash from header matches `MainExtraBodyData` from body. The most appropriate place for it is `Pos.Block.Core.Main.Helpers.verifyBBlock`.

**Comment by rockbmb** [comment]
@7sedam7 All you need to do now is merge `master` and CI will pass.

File: src/Pos/Block/Core/Main/Helpers.hs [hunk 4]
```diff
@@ -36,12 +37,13 @@ instance ( BiHeader ssc
...
          BlockchainHelpers (MainBlockchain ssc) where
     verifyBBlockHeader = verifyMainBlockHeader
-    verifyBBlock UnsafeGenericBlock {..} = do
+    verifyBBlock block@UnsafeGenericBlock {..} = do
         either (throwError . pretty) pure $
             sscVerifyPayload @ssc
                 (Right (Some _gbHeader))
                 (_mbSscPayload _gbBody)
         dlgVerifyPayload (_gbHeader ^. epochIndexL) (_mbDlgPayload _gbBody)
+        unless (hash (block ^. gbExtra) == (block ^. mainBlockEBDataProof)) $ throwError ""
```
Comment (gromakovsky) [inline src/Pos/Block/Core/Main/Helpers.hs]: What is it? Why empty message?

File: src/Pos/Block/Core/Main/Lens.hs [hunk 5]
```diff
@@ -258,6 +264,11 @@ mainBlockHeaderAttributes ::
        Lens' (MainBlock ssc) BlockHeaderAttributes
 mainBlockHeaderAttributes = gbHeader . mainHeaderAttributes
 
+-- | Lens from 'MainBlock' to 'BlockHeaderAttributes'.
```
Comment (gromakovsky) [inline src/Pos/Block/Core/Main/Lens.hs]: Wrong comment.

**Pull request merged**